File: backend/modules/petclinic/app/petclinic/api/owners/createOwner/CreateOwnerService.java
package petclinic.api.owners.createOwner;

import akka.actor.ActorRef;
import akka.pattern.AskSupport;
import akka.pattern.PatternsCS;
import petclinic.api.model.OwnerId;
import petclinic.model.OwnerMessages;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.concurrent.CompletionStage;

import static petclinic.api.Constants.TIMEOUT;

public class CreateOwnerService implements AskSupport {

    @Inject
    @Named("owner")
    private ActorRef owner;

    public CompletionStage<OwnerId> create(String firstName, String lastName, String address, String city, String telephone) {
        return PatternsCS.ask(owner, OwnerMessages.createOwner(firstName, lastName, address, city, telephone), TIMEOUT)
                .thenApply(obj -> OwnerId.apply(obj.toString()));
    }

}

File: backend/modules/petclinic/app/petclinic/api/owners/createOwner/CreateOwnerService.scala
package petclinic.api.owners.addOwner

import akka.actor.ActorRef
import javax.inject.{Inject, Named}
import petclinic.api.Constants
import petclinic.api.model.OwnerId
import petclinic.model.owners.owner.Protocol.CreateOwnerCommand

import scala.concurrent.{ExecutionContext, Future}

class CreateOwnerService @Inject() (@Named(petclinic.api.Constants.OWNER_AGGREGATE_ACTOR_NAME) actorRef: ActorRef)(implicit val ec: ExecutionContext) {

  def create(firstName: String, lastName: String, address: String, city: String, telephone: String): Future[OwnerId] = {
    akka.pattern.Patterns.ask(actorRef, CreateOwnerCommand(firstName, lastName, address, city, telephone), Constants.TIMEOUT)
      .mapTo[String].map(s => OwnerId(s))
  }
}
